### PR TITLE
Refactor auto_route -> go_router

### PR DESCRIPTION
**Pull request by go-finn-nguyen**
## Videos

https://user-images.githubusercontent.com/126144780/233619273-adfc17d0-efd3-4842-b025-b53418719fbf.mov

### CODE REVIEW

File: lib/src/config/constants/navigation_bar_item.dart [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  const XNavigationBarItems({
+    required this.label,
+    required this.route,
+    required this.icon,
+  });
+
+  final String label;
+  final XRouteNames route;
+  final IconData icon;
```
Comment (jimmy-to-goldenowl) [inline lib/src/config/constants/navigation_bar_item.dart:27]: add
`final IconData selectIcon`

File: lib/src/router/coordinator.dart [hunk 2]
```diff
@@ -1,19 +1,52 @@
...
-import 'package:myapp/src/router/auto_router.dart';
-import 'package:myapp/src/router/auto_router.gr.dart';
+import 'package:go_router/go_router.dart';
+import 'package:myapp/src/router/route_name.dart';
+import 'package:myapp/src/router/router.dart';
 
 class XCoordinator {
   static XRouter get rootRouter => GetIt.I<XRouter>();
+  static final navigatorKey = GlobalKey<NavigatorState>();
+  static final shellKey = GlobalKey<NavigatorState>();
```
Comment (jimmy-to-goldenowl) [inline lib/src/router/coordinator.dart:10]: what is different between `shellKey` and `navigatorKey`

File: lib/src/router/coordinator.dart [hunk 3]
```diff
@@ -1,19 +1,52 @@
...
+        extra: extra,
+      );
 
-  static Future showAuthentication() async {
-    return rootRouter.push(const SignRoute());
-  }
+  static void showHomeScreen() => context.goNamed(XRouteNames.home.name);
+  static void showAccountScreen() => context.goNamed(XRouteNames.account.name);
+
+  static Future<T?> showSignInScreen<T extends Object?>() =>
```
Comment (jimmy-to-goldenowl) [inline lib/src/router/coordinator.dart:31]: We should design Navigation stack. and it is root stack.
the stack may like:
```
- Loading,
- Authentication,
                  - Signin
                  - Signup
                  - ForgotPassword
- Dashboard
                  - Home
                  - Profile
                  - Sample
                  - SampleDetail
```
Show for each stack/feature, It has its own Coordinator. for example
lib/src/router/coordinator.dart
 - showDashboard
 - showAuthentication
 
lib/src/feature/Authentication/router/auth_coordinator.dart
 - showSignup
 - showForgotPassword


lib/src/feature/dashboard/router/dashboard_coordinator.dart
 - showSample
 - showSampleDetail

**Review by jimmy-to-goldenowl — CHANGES_REQUESTED** [review]

File: lib/src/router/coordinator.dart (hunk 2 above)
Comment (go-finn-nguyen) [reply lib/src/router/coordinator.dart:10]: `shellKey` maintains the state in `ShellRoute` (nested route), `navigatorKey` like the root key. With those keys, you can specify which screen can have bottom bar, etc.

```dart
final navigatorKey = GlobalKey<NavigatorState>();
final shellKey = GlobalKey<NavigatorState>();

|_ ShellRoute
  |_ GoRoute                            // Needs Bottom Navigation                  
    |_ parentNavigatorKey = shellKey   
  |_ GoRoute                            // Needs Bottom Navigation
    |_ parentNavigatorKey = shellKey   
|_ GoRoute                              // Full Screen which doesn't need Bottom Navigation
  |_parentNavigatorKey = navigatorKey
  ```

File: lib/src/router/coordinator.dart (hunk 3 above)
Comment (go-finn-nguyen) [reply lib/src/router/coordinator.dart:31]: In go_router we can design the navigation stack by defining a route same level or inside another route to depict how the navigation stack will be implemented, you can check this in [router.dart](https://github.com/GoldenOwlAsia/mobile-flutter-template/blob/4c2914d3870e79692bb5bebcfef5ea8755afd889/lib/src/router/router.dart).
Is that what you mean? Or you need to split `coordinator.dart` into multiple files for each feature?

**Pull request merged**